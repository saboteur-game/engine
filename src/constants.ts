export const MAX_PLAYERS = 10;

export const MIN_PLAYERS = 3;

export enum Tools {
  pick = "pick",
  lamp = "lamp",
  wagon = "wagon",
}

interface IRoleRatio {
  [key: number]: {
    saboteurs: number;
    goldDiggers: number;
  };
}

export const roleRatio: IRoleRatio = {
  3: { saboteurs: 1, goldDiggers: 3 },
  4: { saboteurs: 1, goldDiggers: 4 },
  5: { saboteurs: 2, goldDiggers: 4 },
  6: { saboteurs: 2, goldDiggers: 5 },
  7: { saboteurs: 3, goldDiggers: 5 },
  8: { saboteurs: 3, goldDiggers: 6 },
  9: { saboteurs: 3, goldDiggers: 7 },
  10: { saboteurs: 4, goldDiggers: 7 },
};

interface IInitialHandSizes {
  [key: number]: number;
}

export const initialHandSizes: IInitialHandSizes = {
  3: 6,
  4: 6,
  5: 6,
  6: 5,
  7: 5,
  8: 4,
  9: 4,
  10: 4,
};
