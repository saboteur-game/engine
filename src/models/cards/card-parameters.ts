import { Tools } from "../../constants";
import Position from "../position";
import Player from "../player";

export interface IBoardCardParameters {
  position: Position;
}

export interface IToolActionCardParameters {
  player: Player;
  appliedTo: Tools;
}

type CardParameters = IToolActionCardParameters | IBoardCardParameters;

export default CardParameters;
