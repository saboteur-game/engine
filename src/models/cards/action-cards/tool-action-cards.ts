import { Tools } from "../../../constants";
import { ActionCard } from ".";
import { Pojo } from "../../../utils";
import { IToolActionCardParameters } from "../card-parameters";

export class ToolActionCard extends ActionCard {
  appliesTo: Tools[];

  constructor(appliesTo: Tools | Tools[]) {
    super();
    this.appliesTo = ([] as Tools[]).concat(appliesTo);
  }

  play(parameters: IToolActionCardParameters): ActionCard {
    if (!this.appliesTo.includes(parameters.appliedTo)) {
      throw new Error(
        `Invalid parameter provided. Cannot apply to ${parameters.appliedTo}`
      );
    }
    return super.play(parameters);
  }

  toJSON(): Pojo {
    return {
      ...super.toJSON(),
      appliesTo: this.appliesTo,
    };
  }
}

export class BreakActionCard extends ToolActionCard {}

export class RepairActionCard extends ToolActionCard {}
