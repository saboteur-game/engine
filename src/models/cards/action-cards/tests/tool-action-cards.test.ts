import Player from "../../../player";
import { Tools } from "../../../../constants";
import { IToolActionCardParameters } from "../../card-parameters";
import { ToolActionCard } from "../tool-action-cards";

describe("ToolActionCard", () => {
  let toolActionCard: ToolActionCard;
  let cardParameters: IToolActionCardParameters;

  beforeEach(() => {
    toolActionCard = new ToolActionCard(Tools.lamp);
  });

  it("can be serialized", () => {
    expect(toolActionCard.toJSON()).toMatchSnapshot();
  });

  describe("play a card", () => {
    describe("applied to a matching tool", () => {
      beforeEach(() => {
        cardParameters = { player: new Player("Alice"), appliedTo: Tools.lamp };
      });

      it("returns the played card", () => {
        expect(toolActionCard.play(cardParameters)).toBe(toolActionCard);
      });
    });

    describe("applied to a different tool", () => {
      beforeEach(() => {
        cardParameters = { player: new Player("Alice"), appliedTo: Tools.pick };
      });

      it("throws exception", () => {
        expect(() => toolActionCard.play(cardParameters)).toThrow(
          "Invalid parameter provided. Cannot apply to pick"
        );
      });
    });
  });
});
