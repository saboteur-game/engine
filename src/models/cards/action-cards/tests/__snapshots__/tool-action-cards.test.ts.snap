// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolActionCard can be serialized 1`] = `
Object {
  "appliesTo": Array [
    "lamp",
  ],
  "id": "test-id-0",
  "parameters": undefined,
  "status": "unused",
}
`;
